tags/guide/mixinthis.ytag

```diff
@@ -0,0 +1,7 @@
+type: text
+
+---
+
+`(({{0}}) (Object) this)`
+This can be used to access public methods or fields for {{0}} without the use of a `@Shadow` field or method.
+This will not work for private or protected fields or methods. 
```

Comment:
I personally still suggest shadowing than using this ugly code for accessing public stuff. This ugly hack is really created for cases when you need to pass `this` in method calls.

Reply:
Alright, I added that it can be used to pass `this` into methods. I still think it is worth mentioning that it can be used for public methods/fields, as shadow can sometimes be confusing

Reply:
> shadow can sometimes be confusing

How?

Afaik this ugly syntax was created to pass `this` to method calls that require type matches. You shouldn't put calling public stuff as the main purpose of such cast.

Reply:
does this provide a middle ground?:
```
This can be used to pass in `this` to a method that requires a `{{0}}` as a parameter.

A `@Shadow` should be used to access methods and fields of `{{0}}`
```

Reply:
Your "middle ground" is actually exactly what I want :laughing:

But before we make the change, I prefer you to link to examples where people use `(Target) (Object) this` cast to call public methods. I want to check those usages you referred to above and see if they are actually good or bad, like you said "shadows can sometimes be confusing".

Reply:
From fapi:
https://github.com/FabricMC/fabric/blob/c36cc5ad5415c4ef6f7a293cca1bd0ff02589f45/fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/impl/gamerule/rule/BoundedIntRule.java#L63
https://github.com/FabricMC/fabric/blob/c36cc5ad5415c4ef6f7a293cca1bd0ff02589f45/fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/mixin/event/interaction/MixinMinecraftClient.java#L53 (casts to a var and saves it)
https://github.com/FabricMC/fabric/blob/49c073f9a39abebbf1e2a25995118b4418e7a610/fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/mixin/entity/event/LivingEntityMixin.java#L38
https://github.com/FabricMC/fabric/blob/c36cc5ad5415c4ef6f7a293cca1bd0ff02589f45/fabric-game-rule-api-v1/src/main/java/net/fabricmc/fabric/api/gamerule/v1/CustomGameRuleCategory.java#L75
https://github.com/FabricMC/fabric/blob/c36cc5ad5415c4ef6f7a293cca1bd0ff02589f45/fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/mixin/event/interaction/MixinServerPlayerEntity.java#L35
https://github.com/FabricMC/fabric/blob/c36cc5ad5415c4ef6f7a293cca1bd0ff02589f45/fabric-structure-api-v1/src/main/java/net/fabricmc/fabric/api/structure/v1/FabricStructureBuilder.java#L213

just some examples

Reply:
so maybe:
```
`(({{0}}) (Object) this)`
This can be used to pass in `this` to a method that requires a `{{0}}` as a parameter.

A `@Shadow` should be used to access methods and fields of `{{0}}`, but `(({{0}}) (Object) this)` can be used to get access to public methods and fields.
```

Reply:
For bound int, that is an accessor and not in a mixin
for minecraft client, yes that one does reduce declarations in mixin class and can be used as a fine example (reduce the number of double casts as well)
for living entity mixin, i509 actually shadowed the world in the superclass. dunno why he needs to cast to get that world or if mixin doesn't support it
for custom game rule category, also that's not a mixin
for server player entity, that one is okay
for structure builder, that's not a mixin too

-----

(Some of your examples are not double cast of `this`, but double cast to call accessor methods, which is fairly fine)

So after looking at these, the rules I conclude:
For some places where you may shadow too many and stuff is public, feel free to double cast, but try store in local var to reduce code ugliness

otherwise, prefer to shadow to mimic original code (if you don't need to shadow too many)

some stuff, like calling accessors, requires double cast unfortunately.

Feel free to look at my conclusions and edit.